### DISCUSSION TITLE
How to override capabilities at the runtime?

### DISCUSSION THREAD

**Discussion by Shinoni** [webdriverio/webdriverio#10621] [Q&A]
I need to execute some tests against different capabilities, some - against Mobile, some - against browsers within the same test run. I tried to add 'specs' to a capability in wide.conf.js, and, also, tried to override from 'OnStartWorker'/'BeforeSession' hooks. And nothing really worked. Can someone help me if there is a solution for my use case? Thanks.

**Comment by christian-bromann** [comment]
@Shinoni the `wdio.conf.js` can be dynamically generated to allow you setting capabilities based on arbitrary environment conditions, e.g.:
```sh
const config = {
  // ...
}

if (...) {
  config.capabilities = ...
}
```

Note that once a test is started you can't remove a capability anymore.

**Reply by Shinoni** [reply]
My condition is a spec file name. Would that work from wide.conf.js?

**Reply by christian-bromann** [reply]
you can define a custom set of specs in your capabilities, e.g.:

```js
capabilities: [{
    browserName: "chrome"
}, {
    browserName: "firefox",
    specs: [/*specs only for this firefox capability*/]
}]
```

Would this meet your requirements?

**Reply by Shinoni** [reply]
That's exactly what I tried but the top level specs got picked up for execution.

**Reply by christian-bromann** [reply]
You can define it this way:
```js
const specs = [...]

export const config = {
    specs,
    // ...
    capabilities: [{
        browserName: "chrome"
    }, {
        browserName: "firefox",
        specs: [
            ...specs,
            /* additional specs only for this firefox capability */
            '...'
        ]
    }]
    // ...
}
```

**Reply by Shinoni** [reply]
That didn't work either. I'll look further as there might be an issue on our side as we create a custom launcher.

**Reply by Shinoni** [reply]
BTW, while debugging I do see spec files in a capability, so there is something else is going on.

**Reply by Shinoni** [reply]
@christian-bromann Is that expected: when I don't specify 'spec' at the top level, I get 0 spec files.
<img width="1792" alt="Screenshot 2023-06-23 at 3 50 49 PM" src="https://github.com/webdriverio/webdriverio/assets/13089756/a8aa3669-ded2-474f-936c-119fc385103d">

**Reply by Shinoni** [reply]
IsSpecParamPassed is true when this._config.spec = []

**Reply by Shinoni** [reply]
Should we check if it's an array and the array is not empty?

**Reply by Shinoni** [reply]
We create an instance of Launcher and pass an empty array if they're no to-level specs. I fixed the issue on my side but to add a check if an array is not empty won't hurt.
